doc: Add release notes for 2.4 (#11302)
nrfconnect/sdk-nrf: doc/nrf/releases/release-notes-2.4.0.rst

  .. caution::
      This change can affect the Device Firmware Update (DFU) from the older firmware versions that were using the 16-kB settings size.
      Read more about this in the :ref:`ug_matter_device_bootloader_partition_layout` section of the Matter documentation.
      You can still perform DFU from the older firmware version to the latest firmware version, but you will have to change the default settings size from 32 kB to the value used in the older version.

* :ref:`matter_lock_sample` sample:

  * Added the Matter Nordic UART Service (NUS) feature, which allows controlling the door lock device remotely through Bluetooth LE using two simple commands: ``Lock`` and ``Unlock``.
    This feature is dedicated for the nRF52840 and the nRF5340 DKs.
    The sample supports one Bluetooth LE connection at a time.
    Matter commissioning, DFU, and NUS over Bluetooth LE must be run separately.
  * Updated the default value of the feature map.

* :ref:`matter_light_bulb_sample` sample:

  * Added handling of TriggerEffect command.

* :ref:`matter_window_covering_sample` sample:

  * Updated by aligning default values of Window Covering attributes with the specification.

> 

Reviewer: Can we still add one more generic comment in Matter samples section:

Door Lock, Window Covering, Light Bulb and Weather Station applications have been tested with the following ecosystems:
 - Google Home ecosystem for both Matter over Thread and Matter over Wi-Fi solutions. Tested with Google Nest Hub 2nd generation (software version: 47.9.4.447810048; Chromecast firmware version: 1.56.324896, and Google Home mobile application v3.0.1.9).
 - Apple Home ecosystem for both Matter over Thread and Matter over Wi-Fi solutions. Tested with Apple HomePod mini and Apple iPhone (iOS v16.5).
 
Additionally, mentioned samples were positively verified against “Works with Google” certification tests.

Zephyr
======

.. NOTE TO MAINTAINERS: All the Zephyr commits in the below git commands must be handled specially after each upmerge and each nRF Connect SDK release.

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``4bbd91a9083a588002d4397577863e0c54ba7038``, with some |NCS| specific additions.

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

   git log --oneline 4bbd91a908 ^fcaa60a99f

For the list of |NCS| specific commits, including commits cherry-picked from upstream, run:

.. code-block:: none

   git log --oneline manifest-rev ^4bbd91a908

The current |NCS| main branch is based on revision ``4bbd91a908`` of Zephyr.

Additions specific to |NCS|
---------------------------

* Updated TF-M version to 1.7.0.

zcbor
=====

|no_changes_yet_note|

Trusted Firmware-M
==================

|no_changes_yet_note|

> cJSON

Author: @mbolivar-nordic, is there anything to add here?

Reviewer 2: No changes. The cjson repository has not changed at all since v2.3.0.

  * Wi-Fi®:

    * Wi-Fi Station (STA) mode on the nRF5340 host platform.
    * Emulated boards for the nRF7000 and nRF7001 ICs on the nRF7002 :term:`Evaluation Kit (EK)`.
    * Emulated board for the nRF7001 IC on the nRF7002 :term:`Development Kit (DK)`.
    * nRF7002 EB (Expansion Board) for Thingy:53.
    * :ref:`Bluetooth LE coexistence <ug_radio_mpsl_cx_nrf700x>`, a mechanism for managing the Bluetooth LE and Wi-Fi radio operations, so that the two radios operating in close proximity do not cause interference to each other.
    * :ref:`Memfault sample <memfault_sample>` supported on nRF7002 DK, using Wi-Fi for uploading data to the Memfault cloud.

  * Cellular IoT:

    * :ref:`asset_tracker_v2` now supports Wi-Fi locationing using nF9160 DK combined with nRF7002 EK.

  * Security:

    * The Oberon PSA core library provides an optimized implementation of the PSA Crypto API, which reduces the required flash footprint when compared to the Mbed TLS PSA Crypto implementation.

* Added experimental support for the following features:

  * PMIC: nPM1300 PMIC and nPM1300 EK:

>     * Charger, BUCKs, LDOs, Load Switches and GPIOs.

Reviewer 3: ```suggestion
    * Charger, BUCKs, LDOs, Load Switches, and GPIOs.
```

Supported modem firmware
************************

See `Modem firmware compatibility matrix`_ for an overview of which modem firmware versions have been tested with this version of the |NCS|.

Use the latest version of the nRF Programmer app of `nRF Connect for Desktop`_ to update the modem firmware.
See :ref:`nrf9160_gs_updating_fw_modem` for instructions.

Modem-related libraries and versions
====================================

.. list-table:: Modem-related libraries and versions
   :widths: 15 10
   :header-rows: 1

   * - Library name
     - Version information
   * - Modem library
     - `Changelog <Modem library changelog for v2.4.0_>`_
   * - LwM2M carrier library
     - `Changelog <LwM2M carrier library changelog for v2.4.0_>`_

Known issues
************

Known issues are only tracked for the latest official release.
See `known issues for nRF Connect SDK v2.4.0`_ for the list of issues valid for the latest release.

.. _ncs_release_notes_240_changelog:

Changelog
*********

The following sections provide detailed lists of changes by component.

> IDE and tool support

Reviewer 3: Shouldn't we delete sections with the 'no changes yet' note?

Author: Removed all such sections except for Bluetooth LE (waiting for input, if any).

====================

|no_changes_yet_note|

MCUboot
=======

|no_changes_yet_note|

Application development
=======================

|no_changes_yet_note|

RF Front-End Modules
--------------------

|no_changes_yet_note|

Build system
------------

* When using the Kconfig option :kconfig:option:`CONFIG_SB_SIGNING_KEY_FILE` with relative paths, the relative path now points to the application configuration directory instead of the application source directory (these are the same if the application configuration directory is not set).

Security
========

* Updated Mbed TLS version to 3.3.0.

Working with nRF52 Series
=========================

|no_changes_yet_note|

Working with nRF53 Series
=========================

|no_changes_yet_note|

Protocols
=========

This section provides detailed lists of changes by :ref:`protocol <protocols>`.
See `Samples`_ for lists of changes for the protocol-related samples.

Bluetooth LE
------------

|no_changes_yet_note|

Bluetooth mesh
--------------

* Updated:

  * The protocol user guide with the information about the :ref:`dfu_over_bt_mesh`.

>   * The default configuration of advertising sets used by the Bluetooth mesh subsystem to improve performance of the Relay, GATT and Friend features.

Reviewer 3: ```suggestion
  * The default configuration of advertising sets used by the Bluetooth mesh subsystem to improve performance of the Relay, GATT, and Friend features.
```

    This configuration is specified in the :file:`ncs/nrf/subsys/bluetooth/mesh/Kconfig` file.

See `Bluetooth mesh samples`_ for the list of changes in the Bluetooth mesh samples.

Matter
------
* Added:

  * Support for Matter 1.1.0 version.
  * Support for the :ref:`ug_matter_configuring_optional_persistent_subscriptions` feature.
  * Support for negotiating subscription report interval with a Matter Controller to minimalize overall power consumption.
  * The Matter Nordic UART Service (NUS) feature to the :ref:`matter_lock_sample` sample.
    This feature allows using Nordic UART Service to control the device remotely through Bluetooth LE and adding custom text commands to a Matter sample.
    The Matter NUS implementation allows controlling the device regardless of whether the device is connected to a Matter network or not.
    The feature is dedicated for the Matter over Thread solution.
  * Documentation page about :ref:`ug_matter_device_configuring_cd`.
  * Matter SDK fork :ref:`documentation pages <matter_index>` with the page about CHIP Certificate Tool.
  * Documentation page about :ref:`ug_matter_device_adding_bt_services`.

* Updated:

>   * The :ref:`ug_matter` protocol page with a table that lists compatibility versions for the |NCS|, the Matter SDK, and the Matter specification.

Reviewer 3: Suggestion: move doc-related updates to the end of the list

  * The :ref:`ug_matter_tools` page with installation instructions for the ZAP tool, moved from the :ref:`ug_matter_creating_accessory` page.
  * The :ref:`ug_matter_tools` page with information about CHIP Tool, CHIP Certificate Tool, and the Spake2+ Python tool.
  * The :ref:`ug_matter_device_low_power_configuration` page with information about sleepy active threshold parameter configuration and optimizing subscription report intervals.
  * Configuration of factory data partition write-protection on nRF5340 SoC.
  * Configuration of logging verbosity in the debug variants of samples.
  * Overall stability and robustness of Matter over Wi-Fi solution.
  * The default number of user RTC channels on the nRF5340 SoC's network core from 2 to 3 (the platform default) to fix the CSL transmitter feature on Matter over Thread devices acting as Thread routers.
  * :ref:`ug_matter_hw_requirements` with updated memory requirement values valid for the |NCS| v2.4.0.

>   * :ref:`matter_lock_sample`, :ref:`matter_window_covering_sample` and :ref:`matter_light_bulb_sample` samples and the :ref:`matter_weather_station_app` application have been tested with the following ecosystems:

Reviewer 3: ```suggestion
  * :ref:`matter_lock_sample`, :ref:`matter_window_covering_sample`, and :ref:`matter_light_bulb_sample` samples as well as the :ref:`matter_weather_station_app` application have been tested with the following ecosystems:
```

See `known issues for nRF Connect SDK v2.4.0`_ for the list of issues valid for the latest release.

.. _ncs_release_notes_240_changelog:

Changelog
*********

The following sections provide detailed lists of changes by component.

IDE and tool support
====================

|no_changes_yet_note|

MCUboot
=======

|no_changes_yet_note|

Application development
=======================

|no_changes_yet_note|

RF Front-End Modules
--------------------

|no_changes_yet_note|

Build system
------------

* When using :kconfig:option:`CONFIG_SB_SIGNING_KEY_FILE` with relative paths, the relative path is now points to the application configuration directory instead of the application source directory (these are the same if the application configuration directory is not set).

Working with nRF52 Series
=========================

|no_changes_yet_note|

Working with nRF53 Series
=========================

|no_changes_yet_note|

Protocols
=========

This section provides detailed lists of changes by :ref:`protocol <protocols>`.
See `Samples`_ for lists of changes for the protocol-related samples.

Bluetooth® LE
-------------

|no_changes_yet_note|

Bluetooth mesh
--------------

* Updated:

>   * The protocol user guide with the information about :ref:`ble_mesh_dfu_samples`.

Reviewer 4: Might have to change things a bit here. I will know by the end of the day/when PR https://github.com/nrfconnect/sdk-nrf/pull/11209 is completed.

    This feature allows using Nordic UART Service to control the device remotely through Bluetooth LE and adding custom text commands to a Matter sample.
    The Matter NUS implementation allows controlling the device regardless of whether the device is connected to a Matter network or not.
    The feature is dedicated for the Matter over Thread solution.
  * Documentation page about :ref:`ug_matter_device_configuring_cd`.
  * Matter SDK fork :ref:`documentation pages <matter_index>` with the page about CHIP Certificate Tool.
  * Documentation page about :ref:`ug_matter_device_adding_bt_services`.

* Updated:

  * The :ref:`ug_matter` protocol page with a table that lists compatibility versions for the |NCS|, the Matter SDK, and the Matter specification.
  * The :ref:`ug_matter_tools` page with installation instructions for the ZAP tool, moved from the :ref:`ug_matter_creating_accessory` page.
  * The :ref:`ug_matter_tools` page with information about CHIP Tool, CHIP Certificate Tool, and the Spake2+ Python tool.
  * The :ref:`ug_matter_device_low_power_configuration` page with information about sleepy active threshold parameter configuration.

>   * The default number of user RTC channels on the nRF5340 SoC's network core from 2 to 3 (the platform default) to fix the CSL transmitter feature on Matter over Thread devices acting as Thread routers.

Author: Added

See `Matter samples`_ for the list of changes for the Matter samples.

Matter fork
+++++++++++

The Matter fork in the |NCS| (``sdk-connectedhomeip``) contains all commits from the upstream Matter repository up to, and including, the ``v1.1.0.1`` tag.

The following list summarizes the most important changes inherited from the upstream Matter:

* Added:

 * ``SLEEPY_ACTIVE_THRESHOLD`` parameter that makes the Matter sleepy device stay awake for a specified amount of time after network activity.

* Updated:

  * The factory data generation script with the feature for generating the onboarding code.
    You can now use the factory data script to generate a manual pairing code and a QR Code that are required to commission a Matter-enabled device over Bluetooth LE.
    Generated onboarding codes should be put on the device's package or on the device itself.
    For details, see the Generating onboarding codes section on the :doc:`matter:nrfconnect_factory_data_configuration` page in the Matter documentation.
  * The Basic Information cluster with device finish and device color attributes and added the related entries in factory data set.

Thread
------

> |no_changes_yet_note|

Reviewer 5: @peknis 

can we add here a following sentence:

```
Threat in impacted by CVE-2023-2626. Due to this issue, |NCS| v2.4.0 will not undergo the certification process, and is not intended to be used in final Thread products.
```

Reviewer 5: @bpienk fyi

Reviewer 6: Sure. @umapraseeda is in charge here. I don't think we need to link to that known issue (that we added to my PR) as the contents is pretty much the same.

Author: Added this as agreed with @bpienk offline.

See `Thread samples`_ for the list of changes for the Thread samples.

Zigbee
------

* Fixed an issue where device is not fully operational in the distributed network by adding call to the :c:func:`zb_enable_distributed` function for Zigbee Router role by default.

See `Zigbee samples`_ for the list of changes for the Zigbee samples.

> 

Reviewer 7: Could you please add a section for HomeKit:
```suggestion
HomeKit
-------

* Fixed a bug where DFU over UARP protocol does not work on the nRF5340 SoC

```

Author: Added

      The vendor-specific commands for setting the SoC output power and the front-end module gain are not available when the :kconfig:option:`CONFIG_DTM_POWER_CONTROL_AUTOMATIC` Kconfig option is enabled.
    * Support for +1 dBm, +2 dBm, and +3 dBm output power on the nRF5340 DK.

  * Updated the handling of the hardware erratas.
  * Removed a compilation warning when used with minimal pinout Skyworks FEM.

* :ref:`peripheral_uart` sample:

  * Fixed the unit of the :kconfig:option:`CONFIG_BT_NUS_UART_RX_WAIT_TIME` Kconfig option to comply with the UART API.

* :ref:`nrf_dm` sample:

  * Updated the sample by improving the scalability of the sample when it is being used with more devices.

* :ref:`peripheral_fast_pair` sample:

  * Added the default Fast Pair provisioning data that is used when no other provisioning data is specified.
  * Updated the documentation to align it with the new way of displaying notifications for the Fast Pair debug Model IDs.

* Removed the Bluetooth: External radio coexistence using 3-wire interface sample because of the removal of the 3-wire implementation.

Bluetooth mesh samples

> ----------------------

Reviewer 4: PR https://github.com/nrfconnect/sdk-nrf/pull/11273 will affect this section

Author: Added the content under mesh samples section.

* Added:

  * Samples :ref:`ble_mesh_dfu_target` and :ref:`ble_mesh_dfu_distributor` that can be used for evaluation of the Bluetooth mesh DFU specification and subsystem.
  * :ref:`bluetooth_mesh_light_dim` sample that demonstrates how to set up a light dimmer and scene selector application.
* Updated the configuration of advertising sets in all samples to match the new default values.
  See `Bluetooth mesh`_ for more information.
* Removed the :file:`hci_rpmsg.conf` file from all samples that support nRF5340 DK or Thingy:53.
  This configuration is moved to the :file:`ncs/nrf/subsys/bluetooth/mesh/hci_rpmsg_child_image_overlay.conf` file.
* :ref:`bluetooth_mesh_light_lc` sample:

  * Updated to demonstrate the use of sensor server to report additional useful information about the device.

* :ref:`bluetooth_mesh_sensor_server` and :ref:`bluetooth_mesh_sensor_client` samples:

  * Added:

    * Support for motion threshold as a setting for the presence detection.
    * Support for ambient light level sensor.
    * Shell support to :ref:`bluetooth_mesh_sensor_client` sample.

> 

Reviewer 4: ```suggestion

Cryptography samples
--------------------

* Updated all samples and their documentation as the supported logging backend changed to UART instead of both SEGGER RTT and UART.

```

* :ref:`nrf_cloud_rest_fota` sample:

    * Updated the sample by enabling external flash in the nRF9160 DK devicetree overlays for v0.14.0 or later versions, as it is now disabled in the Zephyr board definition.

* :ref:`nrf_cloud_rest_device_message` sample:

  * Added:

    * Overlays to use RTT instead of UART for testing purposes.
    * Sending of log messages directly to nRF Cloud.
    * Overlay to enable `Zephyr Logging`_ backend for full logging to nRF Cloud.

  * Updated the Hello World message sent to nRF Cloud, which now contains a timestamp (message ID).

* :ref:`memfault_sample` sample:

 * Added:
  
    * Support for the nRF7002 DK.
    * A Kconfig fragment to enable ETB trace.

  * Updated:

    * The sample has been moved from :file:`nrf9160/memfault` to :file:`debug/memfault`.
      The documentation is now found in the :ref:`debug_samples` section.

* :ref:`modem_trace_flash` sample:

  * Updated the sample by enabling external flash in the nRF9160 DK devicetree overlays for v0.14.0 or later versions, as it is now disabled in the Zephyr board definition.

Trusted Firmware-M (TF-M) samples
---------------------------------

* :ref:`provisioning_image` sample:

>   * Updated by moving the Thet network core logic to the new sample :ref:`provisioning_image_net_core` instead of being a Zephyr module.

Reviewer 4: Thet?

>   * Updated by moving the Thet network core logic to the new sample :ref:`provisioning_image_net_core` instead of being a Zephyr module.

Reviewer 6: Thet?

Thread samples
--------------

|no_changes_yet_note|

Matter samples
--------------

> * Updated the default settings partition size for all Matter samples from 16 kB to 32 kB.
> 
>   .. caution::
>       This change can affect the Device Firmware Update (DFU) from the older firmware versions that were using the 16-kB settings size.
>       Read more about this in the :ref:`ug_matter_device_bootloader_partition_layout` section of the Matter documentation.
>       You can still perform DFU from the older firmware version to the latest firmware version, but you will have to change the default settings size from 32 kB to the value used in the older version.
> 
> * :ref:`matter_lock_sample` sample:
> 
>   * Added the Matter Nordic UART Service (NUS) feature, which allows controlling the door lock device remotely through Bluetooth LE using two simple commands: ``Lock`` and ``Unlock``.
>     This feature is dedicated for the nRF52840 and the nRF5340 DKs.
>     The sample supports one Bluetooth LE connection at a time.
>     Matter commissioning, DFU, and NUS over Bluetooth LE must be run separately.

Reviewer: Please add the following points:

```suggestion
* Updated the default settings partition size for all Matter samples from 16 kB to 32 kB.

  .. caution::
      This change can affect the Device Firmware Update (DFU) from the older firmware versions that were using the 16-kB settings size.
      Read more about this in the :ref:`ug_matter_device_bootloader_partition_layout` section of the Matter documentation.
      You can still perform DFU from the older firmware version to the latest firmware version, but you will have to change the default settings size from 32 kB to the value used in the older version.

* :ref:`matter_lock_sample` sample:

  * Added the Matter Nordic UART Service (NUS) feature, which allows controlling the door lock device remotely through Bluetooth LE using two simple commands: ``Lock`` and ``Unlock``.
    This feature is dedicated for the nRF52840 and the nRF5340 DKs.
    The sample supports one Bluetooth LE connection at a time.
    Matter commissioning, DFU, and NUS over Bluetooth LE must be run separately.
  * Modified the default value of the feature map.

* :ref:`matter_light_bulb_sample` sample:

  * Added handling of TriggerEffect command.
  
* :ref:`matter_window_covering_sample` sample:

  * Aligned default values of Window Covering attributes with the specification.
```

* :ref:`nrf_modem_lib_readme` library:

  * Added:

    * The :c:func:`nrf_modem_lib_bootloader_init` function to initialize the Modem library in bootloader mode.
    * The function :c:func:`nrf_modem_lib_fault_strerror` to retrieve a statically allocated textual description of a given modem fault.
      The function can be enabled using the new Kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_FAULT_STRERROR`.

  * Updated:

    * The :c:func:`nrf_modem_lib_init` function now initializes the Modem library in normal operating mode only and the ``mode`` parameter is removed.
      Use the :c:func:`nrf_modem_lib_bootloader_init` function to initialize the Modem library in bootloader mode.
    * The Kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_SYS_INIT` is now deprecated.
      The application initializes the modem library using the :c:func:`nrf_modem_lib_init` function instead.
    * The :c:func:`nrf_modem_lib_shutdown` function now checks that the modem is in minimal functional mode (``CFUN=0``) before shutting down the modem.
      If not, a warning is given to the application, and minimal functional mode is set before calling the :c:func:`nrf_modem_shutdown` function.

>     * The Kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_IPC_PRIO_OVERRIDE` is now deprecated.

Reviewer 6: ```suggestion
    * The Kconfig option ``CONFIG_NRF_MODEM_LIB_IPC_PRIO_OVERRIDE`` is now deprecated.
```

      * ``NRF_CLOUD_REST_PGPS_REQ_NO_GPS_TOD``.

    * A-GPS request encoding now uses the common codec function and new nRF Cloud API format.

* :ref:`lib_lwm2m_client_utils` library:

  * Added support for the connection pre-evaluation feature using the Kconfig option :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LTE_CONNEVAL`.
  * Updated the file :file:`lwm2m_client_utils.h`, which includes new API for FOTA to register application callback to receive state changes and requests for the update process.
  * Removed the old API ``lwm2m_firmware_get_update_state_cb()``.

* :ref:`lib_download_client` library:

  * Updated:
  
    * The :c:func:`download_client_connect` function has been refactored to :c:func:`download_client_set_host` and made it non-blocking.
    * The configuration from one security tag to a list of security tags.
    * The library reports error ``ERANGE`` when HTTP range is requested but not supported by server.
  
  * Added:
  
    * The :c:func:`download_client_get` function that combines the functionality of functions :c:func:`download_client_set_host`, :c:func:`download_client_start`, and :c:func:`download_client_disconnect`.
  

>   * Removed functions :c:func:`download_client_pause` and :c:func:`download_client_resume`.

Reviewer 6: ```suggestion
  * Removed functions ``download_client_pause`` and ``download_client_resume``.
```

* :ref:`lib_lwm2m_location_assistance` library:

  * Updated:

    * :file:`lwm2m_client_utils_location.h` includes new API for location assistance to register application callback to receive result codes from location assistance.
    * :file:`lwm2m_client_utils_location.h` by removing deprecated confirmable parameters from location assistance APIs.

* :ref:`pdn_readme` library:

  * Added the event ``PDN_EVENT_NETWORK_DETACH`` to indicate a full network detach.

Libraries for NFC
-----------------

|no_changes_yet_note|

Other libraries
---------------

* Added the :ref:`fem_al_lib` library for use in radio samples.

* :ref:`dk_buttons_and_leds_readme` library:

>   * Updated the library to supports using the GPIO expander for the buttons, switches, and LEDs on the nRF9160 DK.

Reviewer 6: ```suggestion
  * Updated the library to support using the GPIO expander for the buttons, switches, and LEDs on the nRF9160 DK.
```

* :ref:`app_event_manager` library:

  * Added :c:macro:`APP_EVENT_ID` macro.

* :ref:`event_manager_proxy` library:

  * Removed the ``remote_event_name`` argument from the :c:func:`event_manager_proxy_subscribe` function.

* :ref:`mod_memfault`:

  * Added support for the ETB trace to be included in coredump.

Common Application Framework (CAF)
----------------------------------

|no_changes_yet_note|

Shell libraries
---------------

|no_changes_yet_note|

Libraries for Zigbee
--------------------

|no_changes_yet_note|

> sdk-nrfxlib

Reviewer 4: PR https://github.com/nrfconnect/sdk-nrf/pull/11293 will affect this section

Author: Added

  * Added:

    * Support for the device information (``devinfo``) option fetching.
      The option provides device's Vendor ID, Product ID and generation.
    * Support for devices using MCUboot bootloader built in the direct-xip mode (``MCUBOOT+XIP``).
      In this mode, the image is booted directly from the secondary slot without moving it to the primary slot.

MCUboot
=======

The MCUboot fork in |NCS| (``sdk-mcuboot``) contains all commits from the upstream MCUboot repository up to and including ``6902abba270c0fbcbe8ee3bb56fe39bc9acc2774``, with some |NCS| specific additions.

The code for integrating MCUboot into |NCS| is located in the :file:`ncs/nrf/modules/mcuboot` folder.

The following list summarizes both the main changes inherited from upstream MCUboot and the main changes applied to the |NCS| specific additions:

* Added:

  * Support for the downgrade prevention feature using hardware security counters (:kconfig:option:`CONFIG_MCUBOOT_HARDWARE_DOWNGRADE_PREVENTION`).
  * Generation of a new variant of the :file:`dfu_application.zip` when the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option is enabled.

>     Mentioned archive now contains images for both slots, primary and secondary.

Reviewer 6: ```suggestion
    This archive file now contains images for both slots, primary and secondary.
```

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``4bbd91a9083a588002d4397577863e0c54ba7038``, with some |NCS| specific additions.

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

   git log --oneline 4bbd91a908 ^fcaa60a99f

For the list of |NCS| specific commits, including commits cherry-picked from upstream, run:

.. code-block:: none

   git log --oneline manifest-rev ^4bbd91a908

The current |NCS| main branch is based on revision ``4bbd91a908`` of Zephyr.

Additions specific to |NCS|
---------------------------

|no_changes_yet_note|

zcbor
=====

|no_changes_yet_note|

Trusted Firmware-M
==================

|no_changes_yet_note|

cJSON
=====

|no_changes_yet_note|

Documentation
=============

* Added a page on :ref:`ug_nrf70_developing_regulatory_support` in the :ref:`ug_nrf70_developing` user guide.

> 

Reviewer 6: Should we mention here the two new sample categories?

Author: Added

* Updated:

  * The structure of sections on the :ref:`known_issues` page.
    Known issues were moved around, but no changes were made to their description.
    The hardware-only sections were removed and replaced by the "Affected platforms" list.
  * The :ref:`software_maturity` page with details about Bluetooth feature support.
  * The :ref:`ug_nrf5340_gs`, :ref:`ug_thingy53_gs`, :ref:`ug_nrf52_gs`, and :ref:`ug_ble_controller` guides by adding a link to the `Bluetooth LE Fundamentals course`_ in the `Nordic Developer Academy`_.
  * The :ref:`zigbee_weather_station_app` application documentation to match the application template.
  * The :ref:`ug_nrf9160` guide with a section about :ref:`external flash <nrf9160_external_flash>`.
  * The :ref:`nrf_modem_lib_readme` library documentation with a section about :ref:`modem trace flash backend <modem_trace_flash_backend>`.
  * The :ref:`mod_memfault` library documentation has been moved from :ref:`lib_others` to :ref:`lib_debug`.
  * The :ref:`ug_thingy53` guide by adding a section on :ref:`building Wi-Fi applications<thingy53_build_pgm_targets_wifi>` to the .

> 

Reviewer 4: ```suggestion
* The :ref:`ug_bt_mesh_configuring` page with additional configuration options used to configure the behavior and performance of a Bluetooth mesh network.

```

See `Thread samples`_ for the list of changes for the Thread samples.

Zigbee
------

* Fixed an issue where device is not fully operational in the distributed network by adding call to the :c:func:`zb_enable_distributed` function for Zigbee Router role by default.

See `Zigbee samples`_ for the list of changes for the Zigbee samples.

Enhanced ShockBurst (ESB)
-------------------------

* Added:

  * Support for bigger payload size.
    ESB supports a payload with a size of 64 bytes or more.
  * The :ref:`esb_fast_ramp_up` feature that reduces radio ramp-up delay from 130 µs to 40 µs.
  * The :kconfig:option:`CONFIG_ESB_NEVER_DISABLE_TX` Kconfig option as an experimental feature that enables the radio peripheral to remain in TXIDLE state instead of TXDISABLE when transmission is pending.
    For more information, see :ref:`esb_never_disable_tx`.

* Updated:

  * The number of PPI/DPPI channels used from three to six.
  * Events 6 and 7 from the EGU0 instance by assigning them to the ESB module.
  * The type parameter of the :c:func:`esb_set_tx_power` function to ``int8_t``.

nRF IEEE 802.15.4 radio driver
------------------------------

|no_changes_yet_note|

> Wi-Fi

Reviewer 8: I am assuming that this needs to be updated based on board renaming

Reviewer 9: The change is done in this PR and will be integrated here once the PR is merged:
https://github.com/nrfconnect/sdk-nrf/pull/11289/files#diff-17b816d17cd5f06420ac3e251212a721de6dd397df809411ea0b0cf974a877d2

Reviewer 10: Pls add the below update here -

Added
•	Listen interval based power-save – refactored existing power-save configurations into one.
•	Support for configuring antenna gain and band-edge power.
•	Support for nRF700x ns build.
•	Support to interop with CMW500 configured as Wi-Fi Access Point.
•	Support to configure regulatory domain.
•	Support for raw scan results.
Updated:
•	TWT features with improvements (units for interval changed from milliseconds to microseconds).

Reviewer 10: Please add the below updates here -
Added
•	Listen interval based power-save – refactored existing power-save configurations into one.
•	Support for configuring antenna gain and band-edge power.
•	Support for nRF700x ns build.
•	Support to interop with CMW500 configured as Wi-Fi Access Point.
•	Support to configure regulatory domain.
•	Support for raw scan results.
Updated:
•	TWT features with improvements (units for interval changed from milliseconds to microseconds).

Author: Added

* :ref:`peripheral_fast_pair` sample:

  * Added the default Fast Pair provisioning data that is used when no other provisioning data is specified.
  * Updated the documentation to align it with the new way of displaying notifications for the Fast Pair debug Model IDs.

* Removed the Bluetooth: External radio coexistence using 3-wire interface sample because of the removal of the 3-wire implementation.

Bluetooth mesh samples
----------------------

* Added:

  * Samples :ref:`ble_mesh_dfu_target` and :ref:`ble_mesh_dfu_distributor` that can be used for evaluation of the Bluetooth mesh DFU specification and subsystem.
  * :ref:`bluetooth_mesh_light_dim` sample that demonstrates how to set up a light dimmer and scene selector application.
* Updated the configuration of advertising sets in all samples to match the new default values.
  See `Bluetooth mesh`_ for more information.
* Removed the :file:`hci_rpmsg.conf` file from all samples that support nRF5340 DK or Thingy:53.
  This configuration is moved to the :file:`ncs/nrf/subsys/bluetooth/mesh/hci_rpmsg_child_image_overlay.conf` file.
* :ref:`bluetooth_mesh_light_lc` sample:

>   * Updated to demonstrate the use of sensor server to report additional useful information about the device.

Reviewer 11: ```suggestion
  * Updated to demonstrate the use of sensor server to report additional useful information about the device.
  * Added support for the scene server.
```

Author: Added

.. _ncs_release_notes_240_changelog:

Changelog
*********

The following sections provide detailed lists of changes by component.

IDE and tool support
====================

|no_changes_yet_note|

MCUboot
=======

|no_changes_yet_note|

Application development
=======================

|no_changes_yet_note|

RF Front-End Modules
--------------------

|no_changes_yet_note|

Build system
------------

* When using the Kconfig option :kconfig:option:`CONFIG_SB_SIGNING_KEY_FILE` with relative paths, the relative path now points to the application configuration directory instead of the application source directory (these are the same if the application configuration directory is not set).

Working with nRF52 Series
=========================

|no_changes_yet_note|

Working with nRF53 Series
=========================

|no_changes_yet_note|

Protocols
=========

This section provides detailed lists of changes by :ref:`protocol <protocols>`.
See `Samples`_ for lists of changes for the protocol-related samples.

Bluetooth® LE
-------------

|no_changes_yet_note|

Bluetooth mesh
--------------

* Updated:

>   * The protocol user guide with the information about :ref:`ble_mesh_dfu_samples`.

Reviewer 4: ```suggestion
  * The protocol user guide with the information about the :ref:`dfu_over_bt_mesh`.
```

  * Instead of sending a battery voltage, the PMIC's fuel gauge function is used to get a battery percentage.
    For nRF Cloud, the data ID ``VOLTAGE`` has been replaced with ``BATTERY``.
    For the other cloud backends, the name stays the same, but the range changes to 0-100.
  * External flash is enabled in the nRF9160 DK devicetree overlays for v0.14.0 or later versions, as it is now disabled in the Zephyr board definition.

nRF9160: Serial LTE modem
-------------------------

* Added:

  * AT command ``#XWIFIPOS`` to get Wi-Fi location from nRF Cloud.
  * Support for *WRITE REQUEST* in TFTP client.

* Updated the application to use defines from the :ref:`lib_nrf_cloud` library for string values related to nRF Cloud.
* Fixed a bug in receiving a large MQTT Publish message.

nRF5340 Audio
-------------

* Added Kconfig options for setting periodic and extended advertising intervals.
  For more information on the options, see :ref:`nrf53_audio_app_configuration` all options prefixed with ``CONFIG_BLE_ACL_PER_ADV_INT_`` and ``CONFIG_BLE_ACL_EXT_ADV_INT_``.
* Updated:

>   * LE Audio controller for the network core has been moved to the standalone :ref:`lib_bt_ll_acs_nrf53_readme` library.

Author: Added.

  * Updated the handling of the hardware erratas.
  * Removed a compilation warning when used with minimal pinout Skyworks FEM.

* :ref:`peripheral_uart` sample:

  * Fixed the unit of the :kconfig:option:`CONFIG_BT_NUS_UART_RX_WAIT_TIME` Kconfig option to comply with the UART API.

* :ref:`nrf_dm` sample:

  * Updated the sample by improving the scalability when it is used with multiple devices.

* :ref:`peripheral_fast_pair` sample:

  * Added the default Fast Pair provisioning data that is used when no other provisioning data is specified.
  * Updated the documentation to align it with the new way of displaying notifications for the Fast Pair debug Model IDs.

* Removed the Bluetooth: External radio coexistence using 3-wire interface sample because of the removal of the 3-wire implementation.

Bluetooth mesh samples
----------------------

* Added:

  * Samples :ref:`ble_mesh_dfu_target` and :ref:`ble_mesh_dfu_distributor` that can be used for evaluation of the Bluetooth mesh DFU specification and subsystem.
  * :ref:`bluetooth_mesh_light_dim` sample that demonstrates how to set up a light dimmer and scene selector application.

> * Updated the configuration of advertising sets in all samples to match the new default values.
>   See `Bluetooth mesh`_ for more information.

Reviewer 4: ```suggestion
* Updated:

  * The configuration of advertising sets in all samples to match the new default values.
    See `Bluetooth mesh`_ for more information.
  * The link layer payload size over GATT communication from default 27 to maximum 37 bytes (maximum Bluetooth mesh advertiser data) for all samples.
    This was done to avoid fragmentation of outgoing messages to the proxy client.
```

* Removed the :file:`hci_rpmsg.conf` file from all samples that support nRF5340 DK or Thingy:53.
  This configuration is moved to the :file:`ncs/nrf/subsys/bluetooth/mesh/hci_rpmsg_child_image_overlay.conf` file.
* :ref:`bluetooth_mesh_light_lc` sample:

  * Added support for the scene server.
  * Updated to demonstrate the use of sensor server to report additional useful information about the device.

* :ref:`bluetooth_mesh_sensor_server` and :ref:`bluetooth_mesh_sensor_client` samples:

  * Added:

    * Support for motion threshold as a setting for the presence detection.
    * Support for ambient light level sensor.
    * Shell support to :ref:`bluetooth_mesh_sensor_client` sample.

* :ref:`bluetooth_mesh_sensor_server`, :ref:`bluetooth_mesh_light_lc` and :ref:`bluetooth_mesh_light_dim` samples:

> 

Reviewer 4: Would suggest to move this to bullet list in line 388 as suggested above.

    Generated onboarding codes should be put on the device's package or on the device itself.
    For details, see the Generating onboarding codes section on the :doc:`matter:nrfconnect_factory_data_configuration` page in the Matter documentation.
  * The Basic Information cluster with device finish and device color attributes and added the related entries in factory data set.

Thread
------

Thread is impacted by CVE-2023-2626.
Due to this issue, |NCS| v2.4.0 will not undergo the certification process, and is not intended to be used in final Thread products.

Zigbee
------

* Fixed an issue where device is not fully operational in the distributed network by adding a call to the :c:func:`zb_enable_distributed` function for Zigbee Router role by default.

HomeKit
-------

* Fixed a bug where the DFU over UARP protocol does not work on the nRF5340 SoC.

Wi-Fi
-----

* Added support for the following:

  * nRF7000 EK.
  * nRF7001 EK and nRF7001 DK.
  * A new shield, nRF7002 Evaluation Board (EB) for Thingy53 (rev 1.1.0).

>   * Listen interval based power save. Refactored existing power save configurations into one.

Reviewer 3: ```suggestion
  * Listen interval based power save.
     Refactored existing power save configurations into one.
```

    * The :kconfig:option:`CONFIG_BT_FAST_PAIR_STORAGE_USER_RESET_ACTION` Kconfig option to enable a custom user reset action that executes together with the Fast Pair factory reset operation triggered by the :c:func:`bt_fast_pair_factory_reset` function.

  * Updated:

    * The salt size in the Fast Pair not discoverable advertising from 1 byte to 2 bytes, to align with the Fast Pair specification update.
    * The :kconfig:option:`CONFIG_BT_FAST_PAIR_CRYPTO_OBERON` Kconfig option is now the default Fast Pair cryptographic backend.

* :ref:`nrf_bt_scan_readme` library:

  * Fixed the output arguments of the :c:func:`bt_scan_filter_status_get` function.
    The :c:member:`bt_filter_status.manufacturer_data.enabled` field is now correctly set to reflect the status of the filter when the function is called.

Debug libraries
---------------

* Added the :ref:`etb_trace` library for instruction traces.

Modem libraries
---------------

* :ref:`lte_lc_readme` library:

  * Added the Kconfig option :kconfig:option:`CONFIG_LTE_PSM_REQ` that automatically requests PSM on modem initialization.
    If this option is disabled, PSM will not be requested when attaching to the LTE network.

>     This means that the modem's NVS (Non-Volatile Storage) storage contents are ignored.

Reviewer 3: ```suggestion
    This means that the modem's NVS (Non-Volatile Storage) contents are ignored.
```

  * Updated:

    * The Kconfig option :kconfig:option:`CONFIG_LTE_EDRX_REQ`.

>       The option will now prevent the modem from requesting eDRX in case the option is disabled, in contrast to the previous behavior, where eDRX was requested even if the option was disabled (in the case where the modem has preserved requesting eDRX in its NVS storage).

Reviewer 3: ```suggestion
      The option will now prevent the modem from requesting eDRX in case the option is disabled, in contrast to the previous behavior, where eDRX was requested even if the option was disabled (in the case where the modem has preserved requesting eDRX in its non-volatile storage).
```

    * Encode and decode function names have been standardized in the codec.
    * The :c:func:`nrf_cloud_location_request_json_get` function has been moved from the :file:`nrf_cloud_location.h` file to :file:`nrf_cloud_codec.h`.
      The function is now renamed to :c:func:`nrf_cloud_location_request_msg_json_encode`.
    * The library to allow only one file download at a time.
      MQTT-based FOTA, :kconfig:option:`CONFIG_NRF_CLOUD_FOTA`, has priority.

  * Removed unused internal codec function ``nrf_cloud_format_single_cell_pos_req_json()``.

* :ref:`lib_nrf_cloud_rest` library:

  * Updated:

    * The mask angle parameter can now be omitted from an A-GPS REST request by using the value ``NRF_CLOUD_AGPS_MASK_ANGLE_NONE``.
    * The library now uses defines from the :file:`nrf_cloud_pgps.h` file for omitting parameters from a P-GPS request.
      Removed the following values:

      * ``NRF_CLOUD_REST_PGPS_REQ_NO_COUNT``
      * ``NRF_CLOUD_REST_PGPS_REQ_NO_INTERVAL``
      * ``NRF_CLOUD_REST_PGPS_REQ_NO_GPS_DAY``

>       * ``NRF_CLOUD_REST_PGPS_REQ_NO_GPS_TOD``.

Reviewer 3: ```suggestion
      * ``NRF_CLOUD_REST_PGPS_REQ_NO_GPS_TOD``
```

  * Updated:

    * The :c:func:`download_client_connect` function has been refactored to :c:func:`download_client_set_host` and made it non-blocking.
    * The configuration from one security tag to a list of security tags.
    * The library reports error ``ERANGE`` when HTTP range is requested but not supported by server.

  * Removed functions ``download_client_pause()`` and ``download_client_resume()``.

* :ref:`lib_lwm2m_location_assistance` library:

  * Updated:

    * :file:`lwm2m_client_utils_location.h` includes new API for location assistance to register application callback to receive result codes from location assistance.
    * :file:`lwm2m_client_utils_location.h` by removing deprecated confirmable parameters from location assistance APIs.

* :ref:`pdn_readme` library:

  * Added the event ``PDN_EVENT_NETWORK_DETACH`` to indicate a full network detach.

Other libraries
---------------

* Added the :ref:`fem_al_lib` library for use in radio samples.

* :ref:`dk_buttons_and_leds_readme` library:

  * Updated the library to support using the GPIO expander for the buttons, switches, and LEDs on the nRF9160 DK.

* :ref:`app_event_manager` library:

>   * Added :c:macro:`APP_EVENT_ID` macro.

Reviewer 3: ```suggestion
  * Added the :c:macro:`APP_EVENT_ID` macro.
```

The code for integrating MCUboot into |NCS| is located in the :file:`ncs/nrf/modules/mcuboot` folder.

The following list summarizes both the main changes inherited from upstream MCUboot and the main changes applied to the |NCS| specific additions:

* Added:

  * Support for the downgrade prevention feature using hardware security counters (:kconfig:option:`CONFIG_MCUBOOT_HARDWARE_DOWNGRADE_PREVENTION`).
  * Generation of a new variant of the :file:`dfu_application.zip` when the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option is enabled.
    This archive file contains images for both slots, primary and secondary.
  * Encoding of the image start address into the header when the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option is enabled.
    The encoding is done using the ``--rom-fixed`` argument of the :file:`imgtool.py` script.
    If the currently running application also has the :kconfig:option:`CONFIG_MCUMGR_GRP_IMG_REJECT_DIRECT_XIP_MISMATCHED_SLOT` Kconfig option enabled, the MCUmgr rejects application image updates signed without the start address.

> 

Reviewer 12: ```suggestion
* Serial recovery is now able to return hash of image slots, when :kconfig:option:`CONFIG_BOOT_SERIAL_IMG_GRP_HASH` is enabled.

```

Author: Added

Zephyr
======

.. NOTE TO MAINTAINERS: All the Zephyr commits in the below git commands must be handled specially after each upmerge and each nRF Connect SDK release.

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``4bbd91a9083a588002d4397577863e0c54ba7038``, with some |NCS| specific additions.

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

   git log --oneline 4bbd91a908 ^fcaa60a99f

For the list of |NCS| specific commits, including commits cherry-picked from upstream, run:

.. code-block:: none

   git log --oneline manifest-rev ^4bbd91a908

The current |NCS| main branch is based on revision ``4bbd91a908`` of Zephyr.

Additions specific to |NCS|
---------------------------

> * Updated TF-M version to 1.7.0.

Reviewer 13: Shouldn't this be under "Trusted Firmware-M"?

Author: fixed

* :ref:`nrf_dm` sample:

  * Updated the sample by improving the scalability when it is used with multiple devices.

* :ref:`peripheral_fast_pair` sample:

  * Added the default Fast Pair provisioning data that is used when no other provisioning data is specified.
  * Updated the documentation to align it with the new way of displaying notifications for the Fast Pair debug Model IDs.

* Removed the Bluetooth: External radio coexistence using 3-wire interface sample because of the removal of the 3-wire implementation.

Bluetooth mesh samples
----------------------

* Added:

  * Samples :ref:`ble_mesh_dfu_target` and :ref:`ble_mesh_dfu_distributor` that can be used for evaluation of the Bluetooth mesh DFU specification and subsystem.
  * :ref:`bluetooth_mesh_light_dim` sample that demonstrates how to set up a light dimmer and scene selector application.

* Updated:

  * The configuration of advertising sets in all samples to match the new default values.
    See `Bluetooth mesh`_ for more information.
  * The link layer payload size over GATT communication from default 27 to maximum 37 bytes (maximum Bluetooth mesh advertiser data) for all samples.

>     This was done to avoid fragmentation of outgoing messages to the proxy client.

Reviewer 11: Nitpick: Need one extra line break here?

* Removed the :file:`hci_rpmsg.conf` file from all samples that support nRF5340 DK or Thingy:53.
  This configuration is moved to the :file:`ncs/nrf/subsys/bluetooth/mesh/hci_rpmsg_child_image_overlay.conf` file.
* :ref:`bluetooth_mesh_light_lc` sample:

>   * Added support for the scene server.
>   * Updated to demonstrate the use of sensor server to report additional useful information about the device.

Reviewer 4: ```suggestion
  * Added support for the Scene Server model.
  * Updated to demonstrate the use of a Sensor Server model to report additional useful information about the device.
```

    * The Kconfig option ``CONFIG_NRF_MODEM_LIB_SYS_INIT`` is now deprecated.
      The application initializes the modem library using the :c:func:`nrf_modem_lib_init` function instead.
    * The :c:func:`nrf_modem_lib_shutdown` function now checks that the modem is in minimal functional mode (``CFUN=0``) before shutting down the modem.
      If not, a warning is given to the application, and minimal functional mode is set before calling the :c:func:`nrf_modem_shutdown` function.
    * The Kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_IPC_PRIO_OVERRIDE` is now deprecated.

  * Removed:

    * The deprecated function ``nrf_modem_lib_get_init_ret``.
    * The deprecated function ``nrf_modem_lib_shutdown_wait``.
    * The deprecated Kconfig option ``CONFIG_NRF_MODEM_LIB_TRACE_ENABLED``.

* :ref:`pdn_readme` library:

  * Updated the library to use ePCO mode if the Kconfig option :kconfig:option:`CONFIG_PDN_LEGACY_PCO` is not enabled.

  * Fixed:

    * A bug in the initialization of a new PDN context without a PDN event handler.
    * A memory leak in the :c:func:`pdn_ctx_create` function.

> * :ref:`lte_lc_readme` library:

Reviewer 14: LTE LC is present twice, see line 659.

Zephyr
======

.. NOTE TO MAINTAINERS: All the Zephyr commits in the below git commands must be handled specially after each upmerge and each nRF Connect SDK release.

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``4bbd91a9083a588002d4397577863e0c54ba7038``, with some |NCS| specific additions.

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

   git log --oneline 4bbd91a908 ^fcaa60a99f

For the list of |NCS| specific commits, including commits cherry-picked from upstream, run:

.. code-block:: none

   git log --oneline manifest-rev ^4bbd91a908

The current |NCS| main branch is based on revision ``4bbd91a908`` of Zephyr.

Additions specific to |NCS|
---------------------------

|no_changes_yet_note|

zcbor
=====

|no_changes_yet_note|

Trusted Firmware-M
==================

> |no_changes_yet_note|

Reviewer 13: Can you please add

Added: nrf9161 support

Updated: 
TF-M version 1.7.0

Author: Added

    * The Kconfig option ``CONFIG_NRF_MODEM_LIB_SYS_INIT`` is now deprecated.
      The application initializes the modem library using the :c:func:`nrf_modem_lib_init` function instead.
    * The :c:func:`nrf_modem_lib_shutdown` function now checks that the modem is in minimal functional mode (``CFUN=0``) before shutting down the modem.
      If not, a warning is given to the application, and minimal functional mode is set before calling the :c:func:`nrf_modem_shutdown` function.
    * The Kconfig option :kconfig:option:`CONFIG_NRF_MODEM_LIB_IPC_PRIO_OVERRIDE` is now deprecated.

  * Removed:

    * The deprecated function ``nrf_modem_lib_get_init_ret``.
    * The deprecated function ``nrf_modem_lib_shutdown_wait``.
    * The deprecated Kconfig option ``CONFIG_NRF_MODEM_LIB_TRACE_ENABLED``.

* :ref:`pdn_readme` library:

  * Updated the library to use ePCO mode if the Kconfig option :kconfig:option:`CONFIG_PDN_LEGACY_PCO` is not enabled.

  * Fixed:

    * A bug in the initialization of a new PDN context without a PDN event handler.
    * A memory leak in the :c:func:`pdn_ctx_create` function.

* :ref:`lte_lc_readme` library:

>   * Updated:

Reviewer 14: This should also be removed, right?

Author: Yes, removed.

---------------

|no_changes_yet_note|

Libraries for Zigbee
--------------------

|no_changes_yet_note|

sdk-nrfxlib
-----------

* Added:

  * New library :ref:`nrf_fuel_gauge`.
  * PSA core implementation provided by nrf_oberon PSA core.

* Updated:

  * Deprecated PSA core implementation from Mbed TLS.
  * Deprecated PSA driver implementation from Mbed TLS.

See the changelog for each library in the :doc:`nrfxlib documentation <nrfxlib:README>` for additional information.

DFU libraries
-------------

|no_changes_yet_note|

Scripts
=======

This section provides detailed lists of changes by :ref:`script <scripts>`.

* :ref:`partition_manager`:

  * Fixed an issue that prevents an empty gap after a static partition for a region with the ``START_TO_END`` strategy.

* :ref:`nrf_desktop_config_channel_script`:

  * Added:

    * Support for the device information (``devinfo``) option fetching.
      The option provides device's Vendor ID, Product ID and generation.
    * Support for devices using MCUboot bootloader built in the direct-xip mode (``MCUBOOT+XIP``).
      In this mode, the image is booted directly from the secondary slot without moving it to the primary slot.

MCUboot

> =======

Author: @de-nordic , please check.

Author: checked by @de-nordic

The code for integrating MCUboot into |NCS| is located in the :file:`ncs/nrf/modules/mcuboot` folder.

The following list summarizes both the main changes inherited from upstream MCUboot and the main changes applied to the |NCS| specific additions:

* Added:

  * Support for the downgrade prevention feature using hardware security counters (:kconfig:option:`CONFIG_MCUBOOT_HARDWARE_DOWNGRADE_PREVENTION`).
  * Generation of a new variant of the :file:`dfu_application.zip` when the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option is enabled.
    This archive file contains images for both slots, primary and secondary.
  * Encoding of the image start address into the header when the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option is enabled.
    The encoding is done using the ``--rom-fixed`` argument of the :file:`imgtool.py` script.
    If the currently running application also has the :kconfig:option:`CONFIG_MCUMGR_GRP_IMG_REJECT_DIRECT_XIP_MISMATCHED_SLOT` Kconfig option enabled, the MCUmgr rejects application image updates signed without the start address.

Zephyr

> ======

Author: @carlescufi, please check.

Author: Checked

    The encoding is done using the ``--rom-fixed`` argument of the :file:`imgtool.py` script.
    If the currently running application also has the :kconfig:option:`CONFIG_MCUMGR_GRP_IMG_REJECT_DIRECT_XIP_MISMATCHED_SLOT` Kconfig option enabled, the MCUmgr rejects application image updates signed without the start address.

* Updated:

  * Serial recovery is now able to return hash of image slots, when :kconfig:option:`CONFIG_BOOT_SERIAL_IMG_GRP_HASH` is enabled.

Zephyr
======

.. NOTE TO MAINTAINERS: All the Zephyr commits in the below git commands must be handled specially after each upmerge and each nRF Connect SDK release.

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``4bbd91a9083a588002d4397577863e0c54ba7038``, with some |NCS| specific additions.

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

>    git log --oneline 4bbd91a908 ^fcaa60a99f

Reviewer 12: ```suggestion
   git log --oneline 4bbd91a908 ^e1e06d05fa
```
